runtime/connectors/blob/csvreader.go

```diff
@@ -0,0 +1,106 @@
+package blob
+
+import (
+	"bytes"
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"os"
+
+	runtimev1 "github.com/rilldata/rill/proto/gen/rill/runtime/v1"
+	"gocloud.dev/blob"
+)
+
+var _newLineSeparator = []byte("\n")
+
+func downloadCSV(ctx context.Context, bucket *blob.Bucket, obj *blob.ListObject, option *extractOption, fw *os.File) error {
+	reader := NewBlobObjectReader(ctx, bucket, obj)
+
+	rows, err := csvRows(reader, option)
+	if err != nil {
+		return err
+	}
+
+	_, err = fw.Write(rows)
+	return err
+}
+
+func csvRows(reader *ObjectReader, option *extractOption) ([]byte, error) {
+	switch option.strategy {
+	case runtimev1.Source_ExtractPolicy_STRATEGY_HEAD:
+		return csvRowsHead(reader, option)
+	case runtimev1.Source_ExtractPolicy_STRATEGY_TAIL:
+		return csvRowsTail(reader, option)
+	default:
+		panic(fmt.Sprintf("unsupported strategy %s", option.strategy))
+	}
+}
+
+func csvRowsTail(reader *ObjectReader, option *extractOption) ([]byte, error) {
+	header, err := getHeader(reader)
+	if err != nil {
+		return nil, err
+	}
+
+	remBytes := int64(option.limitInBytes - uint64(len(header)))
+	if _, err := reader.Seek(0-remBytes, io.SeekEnd); err != nil {
+		return nil, err
+	}
+
+	p := make([]byte, remBytes)
+	if _, err := reader.Read(p); err != nil {
+		return nil, err
+	}
+
+	lastLineIndex := bytes.Index(p, _newLineSeparator)
+	// remove data before \n since its possibly incomplete
+	// append header at start
+	header = append(header, _newLineSeparator...)
+	return append(header, p[lastLineIndex+1:]...), nil
+}
+
+func csvRowsHead(reader *ObjectReader, option *extractOption) ([]byte, error) {
+	if _, err := reader.Seek(0, io.SeekStart); err != nil {
+		return nil, err
+	}
+
+	p := make([]byte, option.limitInBytes)
+	if _, err := reader.Read(p); err != nil {
+		return nil, err
+	}
```

Comment:
If the file size is smaller than the limit, then won't `reader.Read` succeed, but also return `io.UnexpectedEOF`? 

If the CSV is smaller than the limit, it should just return the full CSV bytes